providers/implementations/encode_decode/ml_kem_codecs.h

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef PROV_ML_KEM_CODECS_H
+# define PROV_ML_KEM_CODECS_H
+# pragma once
+
+# include <openssl/e_os2.h>
+# include "crypto/ml_kem.h"
+# include "prov/provider_ctx.h"
+
+ /*-
+  * The DER ASN.1 encoding of ML-KEM (and ML-DSA) public keys prepends 22 bytes
+  * to the encoded public key:
+  *
+  * - 4 byte outer sequence tag and length
+  * -  2 byte algorithm sequence tag and length
+  * -    2 byte algorithm OID tag and length
+  * -      9 byte algorithm OID (from NIST CSOR OID arc)
+  * -  4 byte bit string tag and length
+  * -    1 bitstring lead byte
+  */
+# define ML_KEM_SPKI_OVERHEAD   22
+typedef struct {
+    const uint8_t asn1_prefix[ML_KEM_SPKI_OVERHEAD];
+} ML_KEM_SPKI_INFO;
+
+/*-
+ * For each parameter set we support a few PKCS#8 input formats, three
+ * corresponding to the "either or both" variants of:
+ *
+ *  ML-KEM-PrivateKey ::= SEQUENCE {
+ *    seed OCTET STRING SIZE (64) OPTIONAL,
+ *    expandedKey [1] IMPLICIT OCTET STRING SIZE (1632 | 2400 | 3172) OPTIONAL }
+ *   (WITH COMPONENTS {..., seed  PRESENT } |
+ *    WITH COMPONENTS {..., expandedKey  PRESENT })
+ *
+ * and two more for historical OQS encodings.
+ *
+ * - OQS private key: OCTET STRING
+ * - OQS private + public key: OCTET STRING
+ *   (The public key is ignored, just as with PKCS#8 v2.)
+ *
+ * An offset of zero means that particular field is absent.
+ */
+typedef struct {
+    const char *p8_name;
+    size_t p8_bytes;
+    uint32_t p8_magic;
```

Comment:
p8_header might be a better name.

Reply:
It was called a magic number by David Hook and others when discussing how such things might be parsed. So not an unexpected term. I'm letting it stand.

Reply:
Just because David chose a poor name doesn't mean we have to perpetuate it.

Reply:
Well I agree, and it isn't exactly a header either.